In [13]:
from sqlalchemy import create_engine
import pandas as pd
engine = create_engine('postgresql://praktikum_student:Sdf4$2;d-d30pp@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru')

In [19]:

hh_salary = pd.read_sql("""
    SELECT 
        ROUND(AVG(salary_from), 2) AS avg_salary_from,
        ROUND(AVG(salary_to), 2) AS avg_salary_to,
        MIN(salary_from) AS min_salary_from,
        MAX(salary_from) AS max_salary_from,
        MIN(salary_to) AS min_salary_to,
        MAX(salary_to) AS max_salary_to
    FROM public.parcing_table
    """, con=engine)


In [21]:
hh_salary

avg_salary_from  avg_salary_to  min_salary_from  max_salary_from  \
0        109525.09      153846.71             50.0         398000.0   

   min_salary_to  max_salary_to  
0        25000.0       497500.0

Средняя зарплата в категории «от» составляет около 109525 рублей, а  
в категории «до» — около 153846 рублей. Это указывает на то, что работодатели готовы платить
аналитикам данных и системным аналитикам в среднем около 130000 рублей. 
Минимальная предлагаемая зарплата начинается с 50 рублей, что, скорее всего, 
является ошибкой данных, а максимальная достигает 497500 рублей.

In [25]:
hh_vacancy = pd.read_sql("""
    SELECT area,
       COUNT(*) AS num_vacancies
    FROM public.parcing_table
    GROUP BY area
    ORDER BY num_vacancies DESC
    """, con=engine)

In [27]:
hh_vacancy

area  num_vacancies
0                      Москва           1247
1             Санкт-Петербург            181
2                Екатеринбург             51
3             Нижний Новгород             33
4                 Новосибирск             33
..                        ...            ...
65                      Псков              1
66                    Иркутск              1
67                  Волгоград              1
68  Северск (Томская область)              1
69                     Курган              1

[70 rows x 2 columns]

Москва и Санкт-Петербург — лидеры по количеству вакансий. 
Это неудивительно, учитывая, что это крупнейшие города 
с развитой инфраструктурой и большим количеством компаний. 
В Екатеринбурге, Нижнем Новгороде и Новосибирске также значительное 
количество вакансий — это указывает на развитый рынок труда для аналитиков данных в этих регионах.

In [29]:
hh_vacancy_by_comp = pd.read_sql("""
    SELECT employer,
       COUNT(*) AS num_vacancies 
    FROM public.parcing_table 
    GROUP BY employer 
    ORDER BY num_vacancies DESC
    """, con=engine)

In [31]:
hh_vacancy_by_comp

employer  num_vacancies
0                 СБЕР            243
1          WILDBERRIES             43
2                 Ozon             34
3       Банк ВТБ (ПАО)             28
4                   Т1             26
..                 ...            ...
685       Бумхиминвест              1
686      Миллениум, ГК              1
687        СЕЛЛ-Сервис              1
688  Контур Компонентс              1
689       Этажи Сибирь              1

[690 rows x 2 columns]

СБЕР предлагает 243 вакансии и является крупнейшим работодателем для аналитиков данных и 
системных аналитиков. Это может говорить о значительных инвестициях в аналитику и технологические решения
в компании. К тому же в банковской среде всегда нужны аналитики. WILDBERRIES, Ozon 
и другие крупные компании также активно ищут специалистов в области данных, 
что говорит о высоком спросе на аналитиков в крупных корпорациях. 

In [33]:
hh_vacancy_by_zanyat = pd.read_sql("""
    SELECT employment,
       COUNT(*) AS num_vacancies
    FROM public.parcing_table 
    GROUP BY employment 
    ORDER BY num_vacancies DESC
    """, con=engine)

In [35]:
hh_vacancy_by_zanyat

employment  num_vacancies
0     Полная занятость           1764
1  Частичная занятость             16
2           Стажировка             16
3     Проектная работа              5

Подавляющее большинство вакансий (1764) предлагают полную занятость.
Это указывает на то, что работодатели предпочитают нанимать 
аналитиков данных и системных аналитиков на постоянные позиции.
Возможно, дело в необходимости глубоко погружаться в проекты и
долго в них участвовать

In [37]:
hh_vacancy_by_schedule = pd.read_sql("""
    SELECT schedule,
       COUNT(*) AS num_vacancies 
    FROM public.parcing_table 
    GROUP BY schedule 
    ORDER BY num_vacancies DESC
    """, con=engine)

In [39]:
hh_vacancy_by_schedule

schedule  num_vacancies
0       Полный день           1441
1  Удаленная работа            310
2     Гибкий график             41
3    Сменный график              9

Большинство вакансий (1441) предлагают работу с полным днём. 
Однако значительное количество вакансий (310) также позволяет 
удалённую работу. Это указывает на то, что работодатели готовы быть
гибкими в современных условиях. Возможно, в ответ на пандемию COVID-19 
и изменения в предпочтениях работников

In [41]:
hh_vacancy_by_grade = pd.read_sql("""
    SELECT experience,
       COUNT(*) AS num_vacancies
    FROM public.parcing_table 
    GROUP BY experience 
    ORDER BY num_vacancies DESC
    """, con=engine)

In [43]:
hh_vacancy_by_grade

experience  num_vacancies
0    Junior+ (1-3 years)           1091
1     Middle (3-6 years)            555
2  Junior (no experince)            142
3      Senior (6+ years)             13

In [ ]:
Наибольшее количество вакансий предназначено для специалистов с опытом от 1 до 3 лет (Junior+), 
что свидетельствует о высоком спросе на специалистов начального и среднего уровней. 
Вакансий для Middle-специалистов (3–6 лет) также много, в то время как спрос 
на Senior-специалистов (6+ лет) крайне низкий, возможно, из-за узкого круга
специалистов с таким уровнем опыта или предпочитаемых долгосрочных позиций.
Однако важно понимать, что Senior-аналитиков могут чаще искать по другим каналам.

In [45]:
hh_hard_soft_skill = pd.read_sql("""
    SELECT key_skills_1,
       COUNT(*) AS num_mention
    FROM public.parcing_table 
    GROUP BY key_skills_1 
    ORDER BY num_mention DESC
    """, con=engine)

In [47]:
hh_hard_soft_skill

key_skills_1  num_mention
0                                     383
1              Анализ данных          312
2                        SQL          161
3               Документация           89
4                     MS SQL           87
..                       ...          ...
134               Qlik Sense            1
135  репутационный маркетинг            1
136              Грамотность            1
137     Экономический анализ            1
138             1С: Торговля            1

[139 rows x 2 columns]

Среди ключевых навыков чаще упоминается «Анализ данных» (312 упоминаний), 
что логично для позиций аналитиков данных. SQL (161 упоминание) и MS SQL (87 упоминаний) также являются
важными навыками, что подчёркивает важность работы с базами данных. «Документация» (89 упоминаний) 
указывает на необходимость ведения точных записей и отчётов. Это особенно актуально для системных аналитиков.
Примечательно, что в топ ключевых навыков не входит Python, однако его часто подразумевают под навыком «Анализ данных».

In [ ]:
import pandas as pd
data = {
    "key": [1, 2, 3, 4, 5, 6],
    "id": [12345, 54321, 98765, 66678, 34567, 34567],
    "phone": ["89997776655", "87778885566", "87776664577", 
              "87778885566", "84547895566", "89087545678"],
    "mail": ["test@mail.ru", "two@mail.ru", "three@mail", 
             "four@mail.ru", "four@mail.ru", "five@mail.ru"]
}

df = pd.DataFrame(data)
df

Для условой таблицы выше необходимо найти все связанные данные

WITH RECURSIVE Relationship(key, id, phone, mail, level) AS (
    SELECT t1.key, t1.id, t1.phone, t1.mail, 1 AS level
    FROM t1
    WHERE phone = '87778885566'  --Тут можно поменять поле и значение 

    UNION

    SELECT t1.key, t1.id, t1.phone, t1.mail, r.level + 1
    FROM t1
    JOIN Relationship r ON (t1.id = r.id 
                      OR t1.phone = r.phone 
                      OR t1.mail = r.mail)
    WHERE r.level < 10 
)
SELECT DISTINCT key, id, phone, mail
FROM Relationship;

Python-версия

In [65]:
import pandas as pd
data = {
    "key": [1, 2, 3, 4, 5, 6],
    "id": [12345, 54321, 98765, 66678, 34567, 34567],
    "phone": ["89997776655", "87778885566", "87776664577", 
              "87778885566", "84547895566", "89087545678"],
    "mail": ["test@mail.ru", "two@mail.ru", "three@mail", 
             "four@mail.ru", "four@mail.ru", "five@mail.ru"]
}

df = pd.DataFrame(data)
df

key     id        phone          mail
0    1  12345  89997776655  test@mail.ru
1    2  54321  87778885566   two@mail.ru
2    3  98765  87776664577    three@mail
3    4  66678  87778885566  four@mail.ru
4    5  34567  84547895566  four@mail.ru
5    6  34567  89087545678  five@mail.ru

In [61]:
def find_related_with_params(df, search_field, search_value, max_level=10):
    # Преобразуем значение в нужный тип данных, если это key
    if search_field in ["id", "key"]:
        search_value = int(search_value)

    relationship = df[df[search_field] == search_value].copy()
    relationship["level"] = 1 

    def recursive_search(related_df, current_level):
        if current_level >= max_level:
            return related_df 

        new_related = df[
            (df["id"].isin(related_df["id"])) |
            (df["phone"].isin(related_df["phone"])) |
            (df["mail"].isin(related_df["mail"]))
        ]

        new_related = new_related[~new_related["key"].isin(related_df["key"])]
        new_related["level"] = current_level + 1

        if new_related.empty:
            return related_df

        combined_df = pd.concat([related_df, new_related])
        return recursive_search(combined_df, current_level + 1)

    return recursive_search(relationship, 1).drop_duplicates()

search_field = input() #Тут вводить поле
search_value = input() #Тут вводить значение 
result_df = find_related_with_params(df, search_field, search_value)

print(result_df[["key", "id", "phone", "mail"]])

 id
 34567


   key     id        phone          mail
4    5  34567  84547895566  four@mail.ru
5    6  34567  89087545678  five@mail.ru
3    4  66678  87778885566  four@mail.ru
1    2  54321  87778885566   two@mail.ru


По таблицам orders и user_actions для каждого дня рассчитайте следующие показатели:

Накопленную выручку на пользователя (Running ARPU).
Накопленную выручку на платящего пользователя (Running ARPPU).
Накопленную выручку с заказа, или средний чек (Running AOV).
Колонки с показателями назовите соответственно running_arpu, running_arppu, running_aov. Колонку с датами назовите dte. 

При расчёте всех показателей округляйте значения до двух знаков после з
пятой.

Результат должен быть отсортирован по возрастанию даты. 

Структура БД:
Таблица Orders:
creation_time
order_id
product_ids - Array, массив хранящий product_id

Таблица user_actions:
action - имеет два состояния 'create_order' и 'cancel_order'
order_id
time
user_id

Таблица Products:
name
price
product_id

SELECT date,
       round(sum(revenue) OVER (ORDER BY date)::decimal / sum(new_users) OVER (ORDER BY date),
             2) as running_arpu,
       round(sum(revenue) OVER (ORDER BY date)::decimal / sum(new_paying_users) OVER (ORDER BY date),
             2) as running_arppu,
       round(sum(revenue) OVER (ORDER BY date)::decimal / sum(orders) OVER (ORDER BY date),
             2) as running_aov
FROM   (SELECT creation_time::date as date,
               count(distinct order_id) as orders,
               sum(price) as revenue
        FROM   (SELECT order_id,
                       creation_time,
                       unnest(product_ids) as product_id
                FROM   orders
                WHERE  order_id not in (SELECT order_id
                                        FROM   user_actions
                                        WHERE  action = 'cancel_order')) t1
            LEFT JOIN products using(product_id)
        GROUP BY date) t2
    LEFT JOIN (SELECT time::date as date,
                      count(distinct user_id) as users
               FROM   user_actions
               GROUP BY date) t3 using (date)
    LEFT JOIN (SELECT time::date as date,
                      count(distinct user_id) as paying_users
               FROM   user_actions
               WHERE  order_id not in (SELECT order_id
                                       FROM   user_actions
                                       WHERE  action = 'cancel_order')
               GROUP BY date) t4 using (date)
    LEFT JOIN (SELECT date,
                      count(user_id) as new_users
               FROM   (SELECT user_id,
                              min(time::date) as date
                       FROM   user_actions
                       GROUP BY user_id) t5
               GROUP BY date) t6 using (date)
    LEFT JOIN (SELECT date,
                      count(user_id) as new_paying_users
               FROM   (SELECT user_id,
                              min(time::date) as date
                       FROM   user_actions
                       WHERE  order_id not in (SELECT order_id
                                               FROM   user_actions
                                               WHERE  action = 'cancel_order')
                       GROUP BY user_id) t7
               GROUP BY date) t8 using (date)